### PR TITLE
[Messenger] Add support for "recording" events from entities

### PR DESCRIPTION
**Pull request by Nyholm**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

This is dependent of #28849. 

```yaml
services:
    messenger.doctrine.entity_message_collector:
        class: Symfony\Bridge\Doctrine\Messenger\EventSubscriber\EntityMessageCollector
        public: false
        arguments: ['@messenger.bus.event']
        tags:
          - { name: 'doctrine.event_subscriber', connection: 'default' }
```

```php

use Symfony\Bridge\Doctrine\EntityMessage\EntityMessageCollectionInterface;
use Symfony\Bridge\Doctrine\EntityMessage\MessageRecorderTrait;

class User implements EntityMessageCollectionInterface
{
    use MessageRecorderTrait;

    // ...
    
    public function setEmail(string $email)
    {
        $oldEmail = $this->email = $email;
        $this->email = $email;
        $this->record(new EmailChanged($this->id, $oldEmail, $email);
    }
```

### CODE REVIEW

**Comment by Koc** [comment]
I'm afraid that this listener should be stateful and events should be dispatched on post flush event https://github.com/doctrine/doctrine2/blob/2.7/lib/Doctrine/ORM/UnitOfWork.php#L430 (because Doctrine's transaction may fall)

**Comment by Nyholm** [comment]
This PR is rebased and squashed

**Review by sroze — CHANGES_REQUESTED** [review]
I agree with @Koc, the `dispatch` of these events needs to happen at `postFlush` otherwise you might dispatch events from rollbacked entities.

But it's a fantastic feature ❤️

**Comment by Nyholm** [comment]
Thank you. I've updated the PR

File: src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Messenger\Stamp\DispatchAfterCurrentBusStamp;
+
+/**
+ * Doctrine listener that listens to Persist, Update and Remove. Every time this is
+ * invoked we take messages from the entities.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ * @author Matthias Noback <matthiasnoback@gmail.com>
```
Comment (Tobion) [inline src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php]: is this based on some existing code? then it should be referenced

Comment (Nyholm) [reply src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php]: How would such reference look?

https://github.com/SimpleBus/SimpleBus/blob/master/Bridge/DoctrineORMBridge/src/EventListener/CollectsEventsFromEntities.php

File: src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        $this->collectEventsFromEntity($event);
+    }
+
+    private function collectEventsFromEntity(LifecycleEventArgs $message)
+    {
+        $entity = $message->getEntity();
+
+        if ($entity instanceof EntityMessageCollectionInterface) {
+            foreach ($entity->getRecordedMessages() as $message) {
+                $this->messageBus->dispatch((new Envelope($message))->with(new DispatchAfterCurrentBusStamp()));
```
Comment (ro0NL) [inline src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php]: shouldwe use the `dispatch($message, [new DispatchAfterCurrentBusStamp()])` sugar in core?

File: src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php (hunk 1 above)
Comment (ogizanagi) [reply src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php]:
```
 * @see https://github.com/SimpleBus/SimpleBus/blob/master/Bridge/DoctrineORMBridge/src/EventListener/CollectsEventsFromEntities.php
```

?

File: src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        $this->collectEventsFromEntity($event);
+    }
+
+    private function collectEventsFromEntity(LifecycleEventArgs $message)
+    {
+        $entity = $message->getEntity();
+
+        if ($entity instanceof EntityMessageCollectionInterface) {
+            foreach ($entity->getRecordedMessages() as $message) {
+                $this->messageBus->dispatch($message, [new DispatchAfterCurrentBusStamp()]);
```
Comment (jakzal) [inline src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php]: What if the message bus throws an exception?

Comment (Nyholm) [reply src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php]: That will bubble up. But the Doctrine transaction will still be committed.

Comment (jakzal) [reply src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php]: So there's a risk some messages will be dispatched while some won't. Not sure if that's an issue yet, just pointing it out.

Comment (Nyholm) [reply src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php]: Yes, If we dispatch 5 messages and number 2 fails, then the 3 other will never be handled. But that is an issue with the message bus and has nothing to do with this PR.

**Comment by Nyholm** [comment]
Maybe @weaverryan could have a look?

**Review by sroze — APPROVED** [review]
Looks good to me, sorry for the delay 👍. 

I imagine you'll PR on `DoctrineBundle` to enable this listener from a configuration flag?

**Comment by sroze** [comment]
@Nyholm can you squad and push again? We should have Travis & AppVeyer rebuilding this :)

**Comment by Nyholm** [comment]
Comments are squashed. I will create a PR to doctrine bundle right away

**Comment by maxhelias** [comment]
Nice feature! This should not be in 4.4 ?

**Comment by vudaltsov** [comment]
What if we avoid leaking events and eliminate the getter?

```
public function dispatch(callable $bus): void
{
    $bus($this->events);
    $this->events = [];
}
```

**Comment by Nyholm** [comment]
Hm. Callables makes the stack tree hard to follow. 

@vudaltsov How would you configure an implementation like this to work with DI?

**Comment by vudaltsov** [comment]
@Nyholm , this would be almost the same code you use:

```php
private function collectEventsFromEntity(LifecycleEventArgs $message): void
{
    $entity = $message->getEntity();

    if (!$entity instanceof EntityMessageCollectionInterface) {
        return;
    }

    $entity->dispatchEvents(function (iterable $events): void {
        foreach ($events as $event) {
            $this->messageBus->dispatch($event, [new DispatchAfterCurrentBusStamp()]);
        }
    });
}
```

**Comment by vudaltsov** [comment]
Also, what if we implement this as a middleware, so that we don't depend on Doctrine? It could do the same thing as the subscriber after the `doctrine_transaction` middleware. It will operate on any object message, not on an `entity`.

File: src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            Events::postFlush,
+        ];
+    }
+
+    public function postFlush(LifecycleEventArgs $event)
+    {
+        $this->collectEventsFromEntity($event);
+    }
+
+    private function collectEventsFromEntity(LifecycleEventArgs $message)
```
Comment (vudaltsov) [inline src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php]: if the method is called `collectEventsFromEntity`, then it should accept `object $entity` as argument?

File: src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    public function getSubscribedEvents()
+    {
+        return [
+            Events::postFlush,
+        ];
+    }
+
+    public function postFlush(LifecycleEventArgs $event)
```
Comment (vudaltsov) [inline src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php]: This code will fail.
The `postFlush` event is not a lifecycle event. It passes an instance of `Doctrine\ORM\Event\PostFlushEventArgs` which is not a subclass of `LifecycleEventArgs`. See https://www.doctrine-project.org/projects/doctrine-orm/en/2.6/reference/events.html#postflush .

**Review by vudaltsov — CHANGES_REQUESTED** [review]

File: src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php (hunk 5 above)
Comment (vudaltsov) [reply src/Symfony/Bridge/Doctrine/Messenger/EventSubscriber/EntityMessageCollector.php]: So this event cannot be used here at all, because at this point you do not know which entities were flushed.

**Comment by Nyholm** [comment]
Im closing this in favor of #34310
